frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/__tests__/utils.spec.ts

```diff
@@ -0,0 +1,49 @@
+import { PipelineExampleNames, pipelineTestData } from '../../../../test-data/pipeline-data';
+import { getLastRegularTasks, getTopologyNodesEdges, getFinallyTaskHeight } from '../utils';
+
+const pipelineData = pipelineTestData[PipelineExampleNames.COMPLEX_PIPELINE];
+const { pipeline } = pipelineData;
+const pipelineWithFinally = {
+  ...pipeline,
+  spec: {
+    ...pipeline.spec,
+    finally: [{ ...pipeline.spec.tasks[0], name: 'finally-task' }],
+  },
+};
+
+describe('getTopologyNodesEdges', () => {
+  it('expect to return nodes and edges for a pipeline without finally tasks', () => {
+    const { nodes, edges } = getTopologyNodesEdges(pipeline);
+    expect(nodes).toHaveLength(pipeline.spec.tasks.length);
+    expect(edges).toHaveLength(19);
+  });
+
+  it('expect to return nodes and edges for a pipeline with finally tasks', () => {
+    const { nodes, edges } = getTopologyNodesEdges(pipelineWithFinally);
+    expect(nodes).toHaveLength(
+      pipelineWithFinally.spec.tasks.length + pipelineWithFinally.spec.finally.length,
+    );
+    expect(edges).toHaveLength(20);
+  });
+});
+
+describe('getLastRegularTasks', () => {
+  it('expect to handle the empty array input', () => {
+    expect(getLastRegularTasks([])).toHaveLength(0);
+  });
+
+  it('expect to return the list of task names that will be executed at the end', () => {
+    const { nodes } = getTopologyNodesEdges(pipeline);
+    expect(getLastRegularTasks(nodes)).toHaveLength(1);
+    expect(getLastRegularTasks(nodes)).toEqual(['verify']);
+  });
```

Comment:
Although a good test, it's not comprehensive. Could we get a test where there are at least two end points?

Reply:
Updated the assertion to test multiple leaf nodes.